File: core-go/changes.go
package main

import (
	"encoding/json"
	"net/http"
	"strconv"
)

// The journal shares state.json's atomic replacement with binding revisions.
// Host writes cannot participate in a database transaction. Prepared/undoing
// entries therefore remain durable and block new writes until reconciled.
type PPTChange struct {
	ID         string           `json:"id"`
	ProjectID  string           `json:"projectId"`
	DocumentID string           `json:"documentId"`
	Label      string           `json:"label"`
	CreatedAt  string           `json:"createdAt"`
	Entries    []PPTChangeEntry `json:"entries"`
}
type PPTChangeEntry struct {
	Target           map[string]any `json:"target"`
	Before           map[string]any `json:"before"`
	After            map[string]any `json:"after,omitempty"`
	Plan             map[string]any `json:"plan"`
	BeforeBinding    *Binding       `json:"beforeBinding"`
	AfterBinding     Binding        `json:"afterBinding"`
	VariableVersion  string         `json:"variableVersion"`
	Status           string         `json:"status"`
	Error            string         `json:"error,omitempty"`
	BindingCommitted bool           `json:"bindingCommitted"`
}

func jsonEqual(a, b any) bool {
	x, _ := json.Marshal(a)
	y, _ := json.Marshal(b)
	return string(x) == string(y)
}
func sameSnapshot(a, b map[string]any) bool {
	if a == nil || b == nil {
		return jsonEqual(a, b)
	}
	if a["comparison"] != nil || b["comparison"] != nil {
		return jsonEqual(a["comparison"], b["comparison"])
	}
	return jsonEqual(a, b)
}
func sameTarget(a, b map[string]any) bool {
	if a["capabilityId"] != nil || b["capabilityId"] != nil {
		return jsonEqual(a["capabilityId"], b["capabilityId"]) && jsonEqual(a["locator"], b["locator"])
	}
	return jsonEqual(a["slideId"], b["slideId"]) && jsonEqual(a["shapeId"], b["shapeId"])
}
func unresolved(c PPTChange) bool {
	for _, e := range c.Entries {
		if e.Status == "prepared" || e.Status == "undoing" {
			return true
		}
	}
	return false
}
func undoable(c PPTChange) bool {
	for _, e := range c.Entries {
		if e.Status == "applied" || e.Status == "prepared" || e.Status == "undoing" {
			return true
		}
	}
	return false
}
func bindingAt(p *Project, id string) *Binding {
	for i := range p.Bindings {
		if p.Bindings[i].ID == id {
			return &p.Bindings[i]
		}
	}
	return nil
}
func (s *Store) ListChanges(pid, doc string) ([]PPTChange, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.projectIndexLocked(pid) < 0 {
		return nil, appErr(404, "项目不存在")
	}
	out := []PPTChange{}
	for i := len(s.State.Changes) - 1; i >= 0; i-- {
		c := s.State.Changes[i]
		if c.ProjectID == pid && c.DocumentID == doc {
			out = append(out, cloneJSON(c))
		}
	}
	return out, nil
}
func (s *Store) PrepareChange(c PPTChange) (PPTChange, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	pi := s.projectIndexLocked(c.ProjectID)
	if pi < 0 {
		return c, appErr(404, "项目不存在")
	}
	p := &s.State.Projects[pi]
	foundDocument := false
	for _, d := range p.Documents {
		foundDocument = foundDocument || d.ID == c.DocumentID
	}
	if !foundDocument {
		return c, appErr(404, "目标文件不属于当前项目")
	}
	for _, old := range s.State.Changes {
		if old.ID == c.ID {
			if old.ProjectID == c.ProjectID && old.DocumentID == c.DocumentID {
				return cloneJSON(old), nil
			}
			return c, appErr(409, "操作编号重复")
		}
		if old.ProjectID == c.ProjectID && old.DocumentID == c.DocumentID && unresolved(old) {
			return c, appErr(409, "有未完成的修改，请先到修改历史中处理")
		}
	}
	if len(c.Entries) == 0 || len(c.Entries) > 200 {
		return c, appErr(400, "修改对象数必须为 1–200")
	}
	for i, e := range c.Entries {
		capability, _ := e.Target["capabilityId"].(string)
		kind, _ := e.Before["kind"].(string)
		adapterSnapshot := capability != "" && e.Before["adapterId"] == capability && kind != ""
		if e.Before == nil || e.Before["version"] != float64(1) || (!adapterSnapshot && kind != "text" && kind != "table") {
			return c, appErr(400, "缺少可恢复的对象快照")
		}
		if e.AfterBinding.DocumentID != c.DocumentID {
			return c, appErr(400, "目标文稿不一致")
		}
		slideID, _ := asFloat(e.Target["slideId"])
		shapeID, _ := asFloat(e.Target["shapeId"])
		locator, _ := e.Target["locator"].(map[string]any)
		if (slideID <= 0 || shapeID <= 0) && (capability == "" || len(locator) == 0) {
			return c, appErr(400, "目标缺少稳定标识")
		}
		for j := 0; j < i; j++ {
			if sameTarget(e.Target, c.Entries[j].Target) {
				return c, appErr(409, "同一批次不能重复修改同一个对象")
			}
		}
		current := bindingAt(p, e.AfterBinding.ID)
		if !jsonEqual(current, e.BeforeBinding) {
			return c, appErr(409, "绑定已变化，请重新生成")
		}
		valid := false
		for _, v := range p.Variables {
			if v.ID == e.AfterBinding.VariableID && v.UpdatedAt == e.VariableVersion {
				valid = true
			}
		}
		if !valid {
			return c, appErr(409, "变量已变化，请重新生成")
		}
		c.Entries[i].Status = "prepared"
		c.Entries[i].After = nil
		c.Entries[i].Error = ""
		c.Entries[i].BindingCommitted = false
	}
	c.CreatedAt = nowISO()
	s.State.Changes = append(s.State.Changes, cloneJSON(c))
	if err := s.saveLocked(); err != nil {
		s.State.Changes = s.State.Changes[:len(s.State.Changes)-1]
		return c, err
	}
	return cloneJSON(c), nil
}
func (s *Store) TransitionChange(pid, id string, index int, action string, snapshot map[string]any, message string) (PPTChange, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	pi := s.projectIndexLocked(pid)
	if pi < 0 {
		return PPTChange{}, appErr(404, "项目不存在")
	}
	ci := -1
	for i, c := range s.State.Changes {
		if c.ID == id && c.ProjectID == pid {
			ci = i
			break
		}
	}
	if ci < 0 {
		return PPTChange{}, appErr(404, "修改记录不存在")
	}
	c := &s.State.Changes[ci]
	if index < 0 || index >= len(c.Entries) {
		return PPTChange{}, appErr(400, "对象序号无效")
	}
	p := &s.State.Projects[pi]
	oldP := cloneJSON(*p)
	oldC := cloneJSON(*c)
	e := &c.Entries[index]
	if action == "undo-start" || action == "recover-start" || action == "undo-complete" {
		for i := len(s.State.Changes) - 1; i > ci; i-- {
			x := s.State.Changes[i]
			if x.ProjectID == pid && x.DocumentID == c.DocumentID && undoable(x) {
				return PPTChange{}, appErr(409, "请先撤销最近一次修改")
			}
		}
		for i := len(c.Entries) - 1; i > index; i-- {
			if c.Entries[i].Status != "undone" && c.Entries[i].Status != "failed" {
				return PPTChange{}, appErr(409, "请按对象修改顺序倒序撤销")
			}
		}
	}
	current := bindingAt(p, e.AfterBinding.ID)
	switch action {
	case "complete":
		for i := 0; i < index; i++ {
			if c.Entries[i].Status == "prepared" || c.Entries[i].Status == "undoing" {
				return PPTChange{}, appErr(409, "前一对象的执行尚未完成")
			}
		}
		if e.Status == "applied" && sameSnapshot(snapshot, e.After) {
			return cloneJSON(*c), nil
		}
		if e.Status != "prepared" || snapshot == nil || snapshot["version"] != float64(1) || snapshot["kind"] != e.Before["kind"] {
			return PPTChange{}, appErr(409, "修改状态不允许完成")
		}
		if !jsonEqual(current, e.BeforeBinding) {
			return PPTChange{}, appErr(409, "绑定已变化，需恢复中断操作")
		}
		if current == nil {
			p.Bindings = append(p.Bindings, cloneJSON(e.AfterBinding))
		} else {
			*current = cloneJSON(e.AfterBinding)
		}
		e.After = cloneJSON(snapshot)
		e.BindingCommitted = true
		e.Status = "applied"
	case "fail":
		if e.Status == "failed" {
			return cloneJSON(*c), nil
		}
		if e.Status != "prepared" || !sameSnapshot(snapshot, e.Before) {
			return PPTChange{}, appErr(409, "未验证恢复原状，不能标记失败")
		}
		e.Status = "failed"
		e.Error = message
	case "undo-start", "recover-start":
		if e.Status == "undoing" {
			return cloneJSON(*c), nil
		}
		if action == "undo-start" && (e.Status != "applied" || (!sameSnapshot(snapshot, e.After) && !sameSnapshot(snapshot, e.Before))) {
			return PPTChange{}, appErr(409, "对象已发生变化，不能自动撤销")
		}
		if action == "recover-start" && e.Status != "prepared" {
			return PPTChange{}, appErr(409, "该操作无需恢复")
		}
		if e.BeforeBinding != nil {
			found := false
			for _, v := range p.Variables {
				if v.ID == e.BeforeBinding.VariableID {
					found = true
				}
			}
			if !found {
				return PPTChange{}, appErr(409, "原绑定引用的变量已删除，无法恢复规则")
			}
		}
		if e.BindingCommitted && !jsonEqual(current, &e.AfterBinding) {
			return PPTChange{}, appErr(409, "绑定规则已变化，不能自动撤销")
		}
		if snapshot == nil {
			return PPTChange{}, appErr(400, "缺少当前状态")
		}
		// Recovery retains the observed partial state for inspection.
		if action == "recover-start" {
			e.After = cloneJSON(snapshot)
		}
		e.Status = "undoing"
	case "undo-complete":
		if e.Status == "undone" {
			return cloneJSON(*c), nil
		}
		if e.Status != "undoing" || !sameSnapshot(snapshot, e.Before) {
			return PPTChange{}, appErr(409, "尚未验证恢复结果")
		}
		if e.BindingCommitted {
			if !jsonEqual(current, &e.AfterBinding) {
				return PPTChange{}, appErr(409, "绑定已变化，恢复结果需检查")
			}
			if e.BeforeBinding != nil {
				*current = cloneJSON(*e.BeforeBinding)
			} else {
				for i, b := range p.Bindings {
					if b.ID == e.AfterBinding.ID {
						p.Bindings = append(p.Bindings[:i], p.Bindings[i+1:]...)
						break
					}
				}
			}
		}
		e.Status = "undone"
	default:
		return PPTChange{}, appErr(400, "未知修改操作")
	}
	p.UpdatedAt = nowISO()
	if err := s.saveLocked(); err != nil {
		*p = oldP
		*c = oldC
		return PPTChange{}, err
	}
	return cloneJSON(*c), nil
}
func (s *Server) changesAPI(w http.ResponseWriter, r *http.Request, pid string, sg []string) {
	if len(sg) == 4 && r.Method == "GET" {
		v, e := s.store.ListChanges(pid, r.URL.Query().Get("documentId"))
		if e != nil {
			s.fail(w, r, e)
			return
		}
		jsonOut(w, r, 200, map[string]any{"changes": v})
		return
	}
	b, e := readJSON(r)
	if e != nil {
		s.fail(w, r, e)
		return
	}
	if len(sg) == 4 && r.Method == "POST" {
		docID, _ := b["documentId"].(string)
		requestID, _ := b["requestId"].(string)
		label, _ := b["label"].(string)
		if requestID == "" {
			s.fail(w, r, appErr(400, "缺少操作编号"))
			return
		}
		// A lost prepare response is retried with the same request ID.
		old, _ := s.store.ListChanges(pid, docID)
		for _, c := range old {
			if c.ID == requestID {
				jsonOut(w, r, 200, map[string]any{"change": c})
				return
			}
		}
		project, err := s.store.GetProject(pid)
		if err != nil {
			s.fail(w, r, err)
			return
		}
		var doc *Document
		for i := range project.Documents {
			if project.Documents[i].ID == docID {
				doc = &project.Documents[i]
			}
		}
		if doc == nil {
			s.fail(w, r, appErr(400, "目标文件 不属于当前项目"))
			return
		}
		c := PPTChange{ID: requestID, ProjectID: pid, DocumentID: docID, Label: label}
		inputs, _ := b["entries"].([]any)
		for _, raw := range inputs {
			in, ok := raw.(map[string]any)
			if !ok {
				s.fail(w, r, appErr(400, "修改条目无效"))
				return
			}
			before, _ := in["before"].(map[string]any)
			draftID, _ := in["draftId"].(string)
			bindingID, _ := in["bindingId"].(string)
			entry := PPTChangeEntry{Before: before}
			if draftID != "" {
				d, err := s.drafts.PeekBinding(pid, draftID)
				if err != nil {
					s.fail(w, r, err)
					return
				}
				approved, _ := in["approveDynamicCapability"].(bool)
				if d.DynamicCapability && !approved {
					s.fail(w, r, appErr(412, "请确认本次临时能力后再应用"))
					return
				}
				if d.DocumentID != docID {
					s.fail(w, r, appErr(409, "草稿不属于当前文稿"))
					return
				}
				entry.Plan = d.Plan
				entry.VariableVersion = d.VariableVersion
				entry.AfterBinding = Binding{ID: newID("bnd"), VariableID: d.VariableID, DocumentID: docID, DocumentKey: doc.Key, Description: d.Description, Target: d.Target, Renderer: d.Renderer, CreatedAt: nowISO(), UpdatedAt: nowISO()}
				if d.BindingID != "" {
					prev, err := s.store.BindingByID(pid, d.BindingID)
					if err != nil {
						s.fail(w, r, err)
						return
					}
					if prev.UpdatedAt != d.BindingVersion {
						s.fail(w, r, appErr(409, "绑定已修改，请重新生成"))
						return
					}
					entry.BeforeBinding = &prev
					entry.AfterBinding.ID = prev.ID
					entry.AfterBinding.CreatedAt = prev.CreatedAt
				}
			} else {
				prev, err := s.store.BindingByID(pid, bindingID)
				if err != nil {
					s.fail(w, r, err)
					return
				}
				entry.BeforeBinding = &prev
				entry.AfterBinding = prev
			}
			v, err := s.store.VariableByID(pid, entry.AfterBinding.VariableID)
			if err != nil {
				s.fail(w, r, err)
				return
			}
			if draftID != "" && entry.VariableVersion != v.UpdatedAt {
				s.fail(w, r, appErr(409, "变量已变化，请重新生成"))
				return
			}
			entry.VariableVersion = v.UpdatedAt
			if draftID == "" {
				entry.Plan, err = RenderPlan(v, entry.AfterBinding.Renderer)
				if err != nil {
					s.fail(w, r, err)
					return
				}
			}
			if !jsonEqual(entry.Plan, in["expectedPlan"]) {
				s.fail(w, r, appErr(409, "渲染结果已变化，请重新生成"))
				return
			}
			entry.Target = entry.AfterBinding.Target
			c.Entries = append(c.Entries, entry)
		}
		result, err := s.store.PrepareChange(c)
		if err != nil {
			s.fail(w, r, err)
			return
		}
		for _, raw := range inputs {
			in, _ := raw.(map[string]any)
			if draftID, _ := in["draftId"].(string); draftID != "" {
				s.drafts.DiscardBinding(pid, draftID)
			}
		}
		jsonOut(w, r, 201, map[string]any{"change": result})
		return
	}
	if len(sg) == 7 && sg[5] == "entries" && r.Method == "POST" {
		index, err := strconv.Atoi(sg[6])
		if err != nil {
			s.fail(w, r, appErr(400, "对象序号无效"))
			return
		}
		action, _ := b["action"].(string)
		snap, _ := b["snapshot"].(map[string]any)
		msg, _ := b["error"].(string)
		result, err := s.store.TransitionChange(pid, sg[4], index, action, snap, msg)
		if err != nil {
			s.fail(w, r, err)
			return
		}
		jsonOut(w, r, 200, map[string]any{"change": result})
		return
	}
	s.fail(w, r, appErr(404, "修改历史接口不存在"))
}

File: agent-host/src/project/changes.ts
import {
  Store,
  entity,
  project,
  id,
  now,
  fingerprint,
  canonical,
} from "./store.js";
import { fail } from "../../../shared/contracts/index.js";
import { render } from "../legacy/execute.js";
const same = (a: any, b: any) => canonical(a ?? null) === canonical(b ?? null);
const sameBinding = (a: any, b: any) => {
  const logical = (value: any) =>
    value
      ? Object.fromEntries(
          Object.entries(value).filter(
            ([key]) => !["revision", "updatedAt"].includes(key),
          ),
        )
      : null;
  return same(logical(a), logical(b));
};
const snapshotSame = (a: any, b: any) =>
  same(a?.comparison || a, b?.comparison || b);
const targetKey = (t: any) =>
  t.capabilityId
    ? { capabilityId: t.capabilityId, locator: t.locator }
    : { slideId: t.slideId, shapeId: t.shapeId };
const unresolved = (c: any) =>
  c.entries.some((e: any) => ["prepared", "undoing"].includes(e.status));
export class Changes {
  constructor(private store: Store) {}
  list(pid: string, documentId: string) {
    return this.store
      .snapshot()
      .pptChanges.filter(
        (c) => c.projectId === pid && c.documentId === documentId,
      )
      .reverse();
  }
  async prepare(pid: string, input: any) {
    return this.store.transaction(async (s) => {
      const p = project(s, pid),
        doc = entity(p.documents, input.documentId);
      if (
        !input.requestId ||
        !Array.isArray(input.entries) ||
        input.entries.length < 1 ||
        input.entries.length > 200
      )
        fail("INVALID_INPUT", "修改请求无效", 400);
      const existing = s.pptChanges.find((c) => c.id === input.requestId);
      if (existing) {
        if (existing.projectId === pid && existing.documentId === doc.id)
          return existing;
        fail("REQUEST_CONFLICT", "操作编号重复", 409);
      }
      if (
        s.pptChanges.some(
          (c) =>
            c.projectId === pid && c.documentId === doc.id && unresolved(c),
        )
      )
        fail("CHANGE_PENDING", "请先恢复未完成修改", 409);
      const entries: any[] = [];
      for (const req of input.entries) {
        if (req.before?.version !== 1 || !req.before.kind)
          fail("INVALID_SNAPSHOT", "缺少可恢复快照", 400);
        let beforeBinding: any = null,
          afterBinding: any,
          plan: any,
          draft: any;
        if (req.draftId) {
          draft = entity(s.drafts, req.draftId);
          if (
            draft.projectId !== pid ||
            draft.kind !== "render" ||
            draft.status !== "preview_ready"
          )
            fail("DRAFT_NOT_READY", "预览未就绪", 409);
          if (draft.requiresRiskAcceptance && !req.acceptRisk)
            fail("RISK_ACK_REQUIRED", "请确认语义复核风险", 412);
          if (draft.input.documentId !== doc.id)
            fail("DOCUMENT_CONFLICT", "目标文稿不一致", 409);
          if (
            draft.input.targetSnapshot &&
            !same(
              fingerprint(req.before?.comparison || req.before),
              draft.targetFingerprint,
            )
          )
            fail("TARGET_CHANGED", "目标对象已经变化，请重新生成预览", 409);
          beforeBinding = draft.bindingId
            ? entity(p.bindings, draft.bindingId)
            : null;
          if (beforeBinding && beforeBinding.revision !== draft.bindingRevision)
            fail("STALE_BINDING_REVISION", "绑定已变化", 409);
          const v = entity(p.variables, draft.variableId);
          if (v.revision !== draft.variableRevision)
            fail("STALE_VARIABLE_REVISION", "变量已变化", 409);
          if (entity(p.sources, v.sourceId).revision !== draft.sourceRevision)
            fail("STALE_SOURCE_REVISION", "源数据已变化", 409);
          afterBinding = {
            id: beforeBinding?.id || id(),
            variableId: v.id,
            documentId: doc.id,
            documentKey: doc.key,
            description: draft.input.description,
            target: draft.input.target,
            renderer: { language: "javascript", version: 2, code: draft.code },
            revision: (beforeBinding?.revision || 0) + 1,
            createdAt: beforeBinding?.createdAt || now(),
            updatedAt: now(),
          };
          plan = draft.result;
        } else {
          beforeBinding = entity(p.bindings, req.bindingId);
          afterBinding = structuredClone(beforeBinding);
          if (beforeBinding.documentId !== doc.id)
            fail("DOCUMENT_CONFLICT", "目标文稿不一致", 409);
          plan = await render(
            entity(p.variables, beforeBinding.variableId),
            beforeBinding.renderer,
            beforeBinding.target,
          );
        }
        if (!same(plan, req.expectedPlan))
          fail("STALE_PLAN", "渲染结果已变化", 409);
        if (
          entries.some((e) =>
            same(targetKey(e.target), targetKey(afterBinding.target)),
          )
        )
          fail("DUPLICATE_TARGET", "不能重复修改同一对象", 409);
        const v = entity(p.variables, afterBinding.variableId);
        entries.push({
          target: afterBinding.target,
          before: req.before,
          plan,
          beforeBinding: beforeBinding ? structuredClone(beforeBinding) : null,
          afterBinding: structuredClone(afterBinding),
          variableVersion: v.updatedAt,
          variableRevision: v.revision,
          status: "prepared",
          bindingCommitted: false,
          draftId: draft?.id,
        });
        if (draft) draft.status = "prepared";
      }
      const change = {
        id: input.requestId,
        projectId: pid,
        documentId: doc.id,
        label: input.label || "更新输出",
        createdAt: now(),
        entries,
      };
      s.pptChanges.push(change);
      return change;
    });
  }
  async transition(pid: string, cid: string, index: number, input: any) {
    const result = await this.store.transaction((s) => {
      const p = project(s, pid),
        c = entity(s.pptChanges, cid);
      if (c.projectId !== pid) fail("NOT_FOUND", "修改记录不存在", 404);
      const e = c.entries[index];
      if (!e) fail("NOT_FOUND", "条目不存在", 404);
      const { action, snapshot } = input,
        current = p.bindings.find((b) => b.id === e.afterBinding.id);
      if (["undo-start", "recover-start", "undo-complete"].includes(action)) {
        const later = s.pptChanges
          .slice(s.pptChanges.indexOf(c) + 1)
          .some(
            (x) =>
              x.projectId === pid &&
              x.documentId === c.documentId &&
              x.entries.some((y: any) =>
                ["applied", "prepared", "undoing"].includes(y.status),
              ),
          );
        if (
          later ||
          c.entries
            .slice(index + 1)
            .some((x: any) => !["undone", "failed"].includes(x.status))
        )
          fail("UNDO_ORDER", "请倒序撤销修改", 409);
      }
      switch (action) {
        case "complete":
          if (e.status === "applied" && snapshotSame(snapshot, e.after))
            return c;
          if (
            c.entries
              .slice(0, index)
              .some((x: any) => ["prepared", "undoing"].includes(x.status))
          )
            fail("CHANGE_ORDER", "前一对象尚未完成", 409);
          if (
            e.status !== "prepared" ||
            snapshot?.version !== 1 ||
            snapshot.kind !== e.before.kind ||
            !same(current, e.beforeBinding)
          )
            fail("CHANGE_CONFLICT", "修改状态或绑定已变化", 409);
          if (
            entity(p.variables, e.afterBinding.variableId).revision !==
            e.variableRevision
          )
            fail(
              "STALE_VARIABLE_REVISION",
              "变量已变化，请恢复本次文档修改",
              409,
            );
          if (current) Object.assign(current, e.afterBinding);
          else p.bindings.push(structuredClone(e.afterBinding));
          e.after = snapshot;
          e.bindingCommitted = true;
          e.status = "applied";
          if (e.draftId) {
            const d = entity(s.drafts, e.draftId);
            e.memoryBeforeScope = (s.memories[d.sessionId] || []).filter(
              (m) => m.scope === `binding:${e.afterBinding.id}`,
            );
            const deltas = d.pendingMemory.map((m: any) => ({
              ...m,
              scope: `binding:${e.afterBinding.id}`,
            }));
            s.memories[d.sessionId] = [
              ...(s.memories[d.sessionId] || []),
              ...deltas,
            ];
            d.pendingMemory = [];
            d.status = "committed";
          }
          break;
        case "fail":
          if (e.status === "failed") return c;
          if (e.status !== "prepared" || !snapshotSame(snapshot, e.before))
            fail("CHANGE_CONFLICT", "未验证恢复原状", 409);
          e.status = "failed";
          e.error = String(input.error || "");
          if (e.draftId) {
            const d = entity(s.drafts, e.draftId);
            d.status = "cancelled";
            d.pendingMemory = [];
          }
          break;
        case "undo-start":
        case "recover-start":
          if (e.status === "undoing") return c;
          if (
            !snapshot ||
            (action === "undo-start" &&
              (e.status !== "applied" ||
                (!snapshotSame(snapshot, e.after) &&
                  !snapshotSame(snapshot, e.before)))) ||
            (action === "recover-start" && e.status !== "prepared")
          )
            fail("TARGET_CHANGED", "对象已变化，不能自动撤销", 409);
          if (e.bindingCommitted && !sameBinding(current, e.afterBinding))
            fail("CHANGE_CONFLICT", "绑定已变化", 409);
          if (e.beforeBinding) entity(p.variables, e.beforeBinding.variableId);
          if (action === "recover-start") e.after = snapshot;
          e.status = "undoing";
          break;
        case "undo-complete":
          if (e.status === "undone") return c;
          if (e.status !== "undoing" || !snapshotSame(snapshot, e.before))
            fail("TARGET_CHANGED", "尚未验证恢复结果", 409);
          if (e.bindingCommitted) {
            if (!sameBinding(current, e.afterBinding))
              fail("CHANGE_CONFLICT", "绑定已变化", 409);
            if (e.beforeBinding)
              Object.assign(current!, e.beforeBinding, {
                revision: current!.revision + 1,
                updatedAt: now(),
              });
            else
              p.bindings = p.bindings.filter((b) => b.id !== e.afterBinding.id);
          }
          e.status = "undone";
          if (e.draftId) {
            const d = entity(s.drafts, e.draftId);
            if (e.bindingCommitted && e.memoryBeforeScope)
              s.memories[d.sessionId] = [
                ...(s.memories[d.sessionId] || []).filter(
                  (m) => m.scope !== `binding:${e.afterBinding.id}`,
                ),
                ...e.memoryBeforeScope,
              ];
            if (d.status !== "committed") {
              d.status = "cancelled";
              d.pendingMemory = [];
            }
          }
          break;
        default:
          fail("INVALID_INPUT", "未知修改操作", 400);
      }
      p.revision++;
      p.updatedAt = now();
      return c;
    });
    await this.store.materializeMemories();
    return result;
  }
}
